[SwiftUI] Scrolling to a particular scroll position does not work when specifying a single axis value

rr-codes opened the pull request:
#### a6f68ea69e73ae1619a17c2a6f4abc4eb8de8105
<pre>
[SwiftUI] Scrolling to a particular scroll position does not work when specifying a single axis value
<a href="https://bugs.webkit.org/show_bug.cgi?id=291161">https://bugs.webkit.org/show_bug.cgi?id=291161</a>
<a href="https://rdar.apple.com/148700838">rdar://148700838</a>

Reviewed by Abrar Rahman Protyasha.

Currently, when using the `.webViewScrollPosition` view modifier on WebView, the following functions
on `ScrollPosition` work properly:

* `.scrollTo(x:y:)`
* `.scrollTo(point:)`
* `.scrollTo(edge:)`

This implements support for the functions that only affect a specific axis:

* `.scrollTo(x:)`
* `.scrollTo(y:)`

so that clients can scroll to a particualr point on one axis and have the other axis unaffected.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::setScrollOffsetWithOptions):
* Source/WebCore/page/LocalFrameView.h:

Implement the main logic to handle reconciling the resolved scroll offset using the specified values,
falling back to the current scroll offset for any axis value that is not specified.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _setContentOffsetX:y:animated:]):
(-[WKWebView _setContentOffset:animated:]): Deleted.

Adjust this function to take seperate optional x and y values in the form of nullable NSNumbers with double
values, which are then converted to `std::optional&lt;int&gt;`s.

* Source/WebKit/UIProcess/Cocoa/WebPageWebView.swift:
(WebPageWebView.setContentOffset(_:y:animated:)):
(WebPageWebView.setContentOffset(_:animated:)): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setContentOffset):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setContentOffset):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/_WebKit_SwiftUI/Implementation/WebViewRepresentable.swift:
(WebViewCoordinator.updateScrollPosition(_:context:)):

Propogate the values from the API to the web process.

Canonical link: <a href="https://commits.webkit.org/293345@main">https://commits.webkit.org/293345@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/26009c106ce973025d7ef9373bd3b04c05997656

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98621 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18254 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8490 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103742 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49201 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100665 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18547 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26705 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75072 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32225 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101625 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14073 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89067 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55430 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13853 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 4 new passes 10 flakes") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7032 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48591 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83808 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7110 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106115 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25711 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18735 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84047 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26088 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85268 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83532 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21101 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28176 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5858 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19401 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25669 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30851 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25487 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28807 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27062 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -2571,6 +2571,24 @@ void LocalFrameView::scrollElementToRect(const Element& element, const IntRect&
 
+void LocalFrameView::setScrollOffsetWithOptions(std::optional<int> x, std::optional<int> y, const ScrollPositionChangeOptions& options)
+{
+    if (x && y) {
+        setScrollOffsetWithOptions({ *x, *y }, options);
+        return;
+    }
```
Optional, but the rest of the function would do this for identical amount of work, right?

rr-codes replied:
it avoids a call to `scrollOffsetFromPosition(scrollPosition())` in this path

aprotyas commented on Source/WebKit/UIProcess/Cocoa/WebPageWebView.swift:
```diff
@@ -100,8 +100,11 @@ extension WebPageWebView {
         set { self._allowsMagnification = newValue }
     }
 
-    public func setContentOffset(_ offset: CGPoint, animated: Bool) {
```
Would it not be convenient to still offer a version that takes an `offset: CGPoint`?

rr-codes replied:
not really, since out of the there call sites of this function, only one actually has a CGPoint, and it's trivial to convert it to its components instead of added code duplication/complexity

rr-codes committed:
[SwiftUI] Scrolling to a particular scroll position does not work when specifying a single axis value
https://bugs.webkit.org/show_bug.cgi?id=291161
rdar://148700838

Reviewed by Abrar Rahman Protyasha.

Currently, when using the `.webViewScrollPosition` view modifier on WebView, the following functions
on `ScrollPosition` work properly:

* `.scrollTo(x:y:)`
* `.scrollTo(point:)`
* `.scrollTo(edge:)`

This implements support for the functions that only affect a specific axis:

* `.scrollTo(x:)`
* `.scrollTo(y:)`

so that clients can scroll to a particualr point on one axis and have the other axis unaffected.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::setScrollOffsetWithOptions):
* Source/WebCore/page/LocalFrameView.h:

Implement the main logic to handle reconciling the resolved scroll offset using the specified values,
falling back to the current scroll offset for any axis value that is not specified.

* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _setContentOffsetX:y:animated:]):
(-[WKWebView _setContentOffset:animated:]): Deleted.

Adjust this function to take seperate optional x and y values in the form of nullable NSNumbers with double
values, which are then converted to `std::optional<int>`s.

* Source/WebKit/UIProcess/Cocoa/WebPageWebView.swift:
(WebPageWebView.setContentOffset(_:y:animated:)):
(WebPageWebView.setContentOffset(_:animated:)): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setContentOffset):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setContentOffset):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/_WebKit_SwiftUI/Implementation/WebViewRepresentable.swift:
(WebViewCoordinator.updateScrollPosition(_:context:)):

Propogate the values from the API to the web process.

Canonical link: https://commits.webkit.org/293345@main
